### ISSUE TITLE
Support Parsing Nullish Coalescing, Optional Chaining and Assertion Functions (TS 3.7)

### ISSUE THREAD

**Issue by c-bandy** [typescript-eslint/typescript-eslint#1033]
The [TypeScript 3.7 beta](https://devblogs.microsoft.com/typescript/announcing-typescript-3-7-beta/) is out.

There should be support to lint code using `typescript-eslint` with the new features in this version: nullish coalescing, optional chaining and assertion functions! 

_This post was changed from it's original state as it's pinned now ❤️_

**Comment by a-tarasyuk** [comment]
@c-bandy Thanks for the report. `typescript-eslint` doesn't currently support the latest version (`3.7.0-beta`) of TypeScript. The new version was published yesterday and includes several [new features](https://devblogs.microsoft.com/typescript/announcing-typescript-3-7-beta/). After the 3.7.0 will be released, definitely typescript-eslint will support it.

Also, `typescript-eslint` reports about that 

```
WARNING: You are currently running a version of TypeScript which is not officially supported by typescript-estree.
You may find that it works just fine, or you may not.
SUPPORTED TYPESCRIPT VERSIONS: >=3.2.1 <3.7.0
YOUR TYPESCRIPT VERSION: 3.7.0-beta
Please only submit bug reports when using the officially supported version.
```

cc @bradzacher

**Comment by c-bandy** [comment]
@a-tarasyuk I'm aware it's not officially supported. It seemed like it was _at least_ partially supported because I could get optional chaining to lint without any problems (`foo?.bar`). I'm not very experienced in this area so I don't know if it was just coincidence or not, but thanks for your input! 🙂

**Comment by a-tarasyuk** [comment]
@c-bandy 

>  It seemed like it was at least partially supported because I could get optional chaining (foo?.bar) to 
lint without any problems.

Based on the release notes _optional chaining?_ was implemented in `3.7.0-beta`.

**Comment by narthollis** [comment]
3.7.0-beta includes both `Nullish Coalescing` and `Optional Chaining` (nullish coalescing is the next major heading after optional chaining in the release notes)

I am haply using both of these features with typescript 3.7.0-beta, except that eslint and other tools that use typescript-estree are giving errors when they run into `??`.

**Comment by c-bandy** [comment]
@a-tarasyuk That's how I knew about it to begin with, I was just surprised optional chaining worked out of the box

**Comment by bradzacher** [comment]
My dudes! You're too quick! The beta was only released yesterday.

We'll ofc be adding support for it. Because it's new syntax, we'll have to build in support to the parser for it.

Optional chaining probably emits an extra prop onto an existing node - hence it's "supported" (note that it's not actually supported, it just doesn't throw an error).
Nullish coalescing, on the other hand, is a new operator that will require explicit support to not throw errors.

**Comment by AlCalzone** [comment]
> Optional chaining probably emits an extra prop onto an existing node - hence it's "supported" (note that it's not actually supported, it just doesn't throw an error).

So is this
![grafik](https://user-images.githubusercontent.com/17641229/66072585-9b2e6280-e555-11e9-8d7d-9cf583ad1ad0.png)
an error in eslint or the prettier rule?

**Comment by bradzacher** [comment]
uhhh - I don't actually know. It parses fine using our parser.
But the prettier plugin defers to our parser as well... so it should just work?
I'm not sure why that happens!

**Comment by bradzacher** [comment]
I used a local version of astexplorer to test this with typescript@beta.

Optional chaining:

```ts
x?.concat([]);
```

<details>
<summary>TS provides the following AST</summary>

```jsonc
{
  "pos": 0,
  "end": 15,
  "flags": 0,
  "modifierFlagsCache": 0,
  "transformFlags": 0,
  "kind": 288,
  "text": "x?.concat([]);\n",
  "bindDiagnostics": [],
  "languageVersion": 99,
  "fileName": "astExplorer.tsx",
  "languageVariant": 1,
  "isDeclarationFile": false,
  "scriptKind": 4,
  "pragmas": {},
  "referencedFiles": [],
  "typeReferenceDirectives": [],
  "libReferenceDirectives": [],
  "amdDependencies": [],
  "hasNoDefaultLib": false,
  "statements": [
    {
      "pos": 0,
      "end": 14,
      "flags": 0,
      "modifierFlagsCache": 536870912,
      "transformFlags": 0,
      "parent": "[Circular ~]",
      "kind": 225,
      "expression": {
        "pos": 0,
        "end": 13,
        "flags": 32,
        "modifierFlagsCache": 0,
        "transformFlags": 0,
        "parent": "[Circular ~.statements.0]",
        "kind": 195,
        "expression": {
          "pos": 0,
          "end": 9,
          "flags": 32,
          "modifierFlagsCache": 0,
          "transformFlags": 0,
          "parent": "[Circular ~.statements.0.expression]",
          "kind": 193,
          "expression": {
            "pos": 0,
            "end": 1,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "parent": "[Circular ~.statements.0.expression.expression]",
            "kind": 75,
            "escapedText": "x"
          },
//////////// THE NEW "QuestionDotToken" ////////////
          "questionDotToken": {
            "pos": 1,
            "end": 3,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "parent": "[Circular ~.statements.0.expression.expression]",
            "kind": 28
          },
////////////////////////////////////////////////////
          "name": {
            "pos": 3,
            "end": 9,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "parent": "[Circular ~.statements.0.expression.expression]",
            "kind": 75,
            "escapedText": "concat"
          }
        },
        "arguments": [
          {
            "pos": 10,
            "end": 12,
            "flags": 0,
            "modifierFlagsCache": 0,
            "transformFlags": 0,
            "parent": "[Circular ~.statements.0.expression]",
            "kind": 191,
            "elements": []
          }
        ]
      }
    }
  ],
  "endOfFileToken": {
    "pos": 14,
    "end": 15,
    "flags": 0,
    "modifierFlagsCache": 0,
    "transformFlags": 0,
    "parent": "[Circular ~]",
    "kind": 1
  },
  "nodeCount": 9,
  "identifierCount": 2,
  "identifiers": {},
  "parseDiagnostics": [],
  "path": "astExplorer.tsx",
  "resolvedPath": "astExplorer.tsx",
  "originalFileName": "astExplorer.tsx",
  "imports": [],
  "moduleAugmentations": [],
  "ambientModuleNames": []
}
```

</details>

---

Nullish coalescing

```ts
x ?? y
```

<details>
<summary>TS provides the following AST</summary>

```jsonc
{
  "pos": 0,
  "end": 8,
  "flags": 0,
  "modifierFlagsCache": 0,
  "transformFlags": 0,
  "kind": 288,
  "text": "x ?? y;\n",
  "bindDiagnostics": [],
  "languageVersion": 99,
  "fileName": "astExplorer.tsx",
  "languageVariant": 1,
  "isDeclarationFile": false,
  "scriptKind": 4,
  "pragmas": {},
  "referencedFiles": [],
  "typeReferenceDirectives": [],
  "libReferenceDirectives": [],
  "amdDependencies": [],
  "hasNoDefaultLib": false,
  "statements": [
    {
      "pos": 0,
      "end": 7,
      "flags": 0,
      "modifierFlagsCache": 536870912,
      "transformFlags": 0,
      "parent": "[Circular ~]",
      "kind": 225,
      "expression": {
        "pos": 0,
        "end": 6,
        "flags": 0,
        "modifierFlagsCache": 0,
        "transformFlags": 0,
        "parent": "[Circular ~.statements.0]",
        "kind": 208,
        "left": {
          "pos": 0,
          "end": 1,
          "flags": 0,
          "modifierFlagsCache": 0,
          "transformFlags": 0,
          "parent": "[Circular ~.statements.0.expression]",
          "kind": 75,
          "escapedText": "x"
        },
//////////// THE NEW "QuestionQuestionToken" ////////////
        "operatorToken": {
          "pos": 1,
          "end": 4,
          "flags": 0,
          "modifierFlagsCache": 0,
          "transformFlags": 0,
          "parent": "[Circular ~.statements.0.expression]",
          "kind": 60
        },
/////////////////////////////////////////////////////////
        "right": {
          "pos": 4,
          "end": 6,
          "flags": 0,
          "modifierFlagsCache": 0,
          "transformFlags": 0,
          "parent": "[Circular ~.statements.0.expression]",
          "kind": 75,
          "escapedText": "y"
        }
      }
    }
  ],
  "endOfFileToken": {
    "pos": 7,
    "end": 8,
    "flags": 0,
    "modifierFlagsCache": 0,
    "transformFlags": 0,
    "parent": "[Circular ~]",
    "kind": 1
  },
  "nodeCount": 7,
  "identifierCount": 2,
  "identifiers": {},
  "parseDiagnostics": [],
  "path": "astExplorer.tsx",
  "resolvedPath": "astExplorer.tsx",
  "originalFileName": "astExplorer.tsx",
  "imports": [],
  "moduleAugmentations": [],
  "ambientModuleNames": []
}
```

</details>

**Comment by bytenik** [comment]
> Thinking about it, it'll probably be a larger piece of work to add support for this across all of the rules, but at least I can stop the parser barfing for you.

Yes, "anti-barfing support" would be fantastic. 😎 Real support is less critical than being able to continue to use existing eslint functionality while this sort of thing gets worked out. Thanks a bunch!

**Comment by narthollis** [comment]
> My dudes! You're too quick! The beta was only released yesterday.
>

This is true - I was mostly just looking for a tracking item - apologies for the noise.

> Thinking about it, it'll probably be a larger piece of work to add support for this across all of the rules, but at least I can stop the parser barfing for you.

This would be fantastic - being able to keep using tools even without direct support would great.

**Comment by bradzacher** [comment]
Just updating - this issue is specifically for parsing.
For supporting the syntax in rules, see #1051 and #1052

**Comment by bradzacher** [comment]
Official support has been released!

Note that there shouldn't have been any parser errors on these, the errors people were getting were because prettier hard errors on things it doesn't support.

If you want to track progress in prettier, see
https://github.com/prettier/prettier/issues/6595
https://github.com/prettier/prettier/issues/6609

**Comment by RichieRunner** [comment]
> Official support has been released!
> 
> Note that there shouldn't have been any parser errors on these, the errors people were getting were because prettier hard errors on things it doesn't support.
> 
> If you want to track progress in prettier, see
> [prettier/prettier#6595](https://github.com/prettier/prettier/issues/6595)
> [prettier/prettier#6609](https://github.com/prettier/prettier/issues/6609)

Which release is this in?

**Comment by bradzacher** [comment]
https://github.com/typescript-eslint/typescript-eslint/blob/master/CHANGELOG.md

**Comment by damiangreen** [comment]
I've been struggling to get eslint to understand  typescript's 3.7 features, especially optional chaining. I have  `parser: '@typescript-eslint/parser',`
`   '@typescript-eslint',` in my plugins.

however I get `undef` errors when running eslint.  
Would be amazing if someone can shed some light on what piece of the puzzle i'm missing

**Comment by bradzacher** [comment]
If you're looking for support in the rules, see the issues linked in
https://github.com/typescript-eslint/typescript-eslint/issues/1033#issuecomment-539311180

As mentioned in the title, and in the comment above - this issue is specifically about parsing support (i.e. making sure that an AST is produced so that eslint won't crash out on the new syntax).


There are a number of issues with some of the base rules due to how eslint works.
These are logged and PRs have been raised. #1090 #1104 #1116 #1138

For everyone:
If you'd like to see wider support for the new syntax in the rules, feel free to hop in and raise some PRs.
We are a community run project, so we rely on community contributions to help us move fast.

**Comment by kalbert312** [comment]
A note on the `ESLint parsing error: Expression expected`... if your IDE is running ESLint in the background and you upgrade your packages you need to restart the IDE or the ESLint process. Source: also ran into this.
